### ISSUE TITLE
OpenGL version on OSX 10.10

### ISSUE THREAD

**Issue by sturmE** [RPCS3/rpcs3#1028]
In trying to run the convert_swizzle sample on OSX 10.10, I get 

```
RSX: E {RSXThread} Failed to compile shader: ERROR: 0:1: '' :  version '420' is not supported
```

Followed by ```INVALID_OPERATION``` failures on most OpenGL calls. I added some code to check out my GL and GLSL version.

```
printf("gl: %s\n", glGetString(GL_VERSION));
printf("glsl: %s\n", glGetString(GL_SHADING_LANGUAGE_VERSION));
```

Which gives me 

```
gl: 2.1 INTEL-10.2.46
glsl: 1.20
```

I'm pretty sure thats the problem. OpenGL version 2.1 with shader language 120. 

I looked into how the OpenGL context is created and it looks like it is done by the wxWidgets library when a ```wxGLCanvas``` is created in ```GLGSFrame::GLGSFrame()```. Looking into the wxWigdets documentation it looks like you can control the version by passing in some values to the ```wxGLCanvas``` constructor 

http://docs.wxwidgets.org/trunk/glcanvas_8h.html#aae05225933a42f81e7c4a9fb286596f9aa75a4e738b4fa60c91409090f40d7a40

But it seems that the version of wxWidgets used in this project does not support this.

Does anyone have a solution or ideas related to this issue?

**Comment by tambry** [comment]
@sturmE This is error is caused by you having an old video card, which doesn't support OpenGL 4.2 shading language. To fix this error replace your video card with a newer one, which supports OpenGL 4.2 shading language.

**Comment by sturmE** [comment]
False. That is not the case. Here is me running another opengl program I wrote.

```
(gfx [DEBUG]) 4.1 INTEL-10.2.46
(gfx [DEBUG]) 4.10
```

**Comment by tambry** [comment]
@sturmE Do you have an (integrated) Intel graphics card? Still seems like your graphics card only supports OpenGL 4.1, while OpenGL 4.2 is required.

**Comment by sturmE** [comment]
@tambry Yes it is, but OSX doesn't support > 4.1. I assume [this](https://github.com/DHrpcs3/rpcs3/commit/2c98a37512731f93a94b0700e4f1a9e6982b8e5b) is the reason 4.2 is required? According to that issue it is because of a call to [packSNorm4x8](https://www.opengl.org/sdk/docs/man/html/packUnorm.xhtml) which is actually supported in 4.1. So even if 4.2 is required, I'm pretty sure 4.1 will also suffice (at least at this point in time). The problem, and reason for this question, is that  whoever is responsible for creating the OpengGL context, which seemss to be happing inside wxWidgets, is not doing it correctly for OSX and creating a 2.1 instead of a 4.1 context. I was curious if other people knew about this, had a fix for this or any other relevant information on the topic.

**Comment by Syphurith** [comment]
@sturmE well you can try change the 420 to 410 to see if that compiles and produces result. I should warn you of any rendering problems after that.. BTW your graphic card seems to be an Intel one. If that is not your primary card, find some steps to select the other card for rpcs3 may easily solve this.

**Comment by sturmE** [comment]
@Syphurith Thanks. 

I think the core question I am asking still isn't getting answered due to me not being specific enough.

What I am trying to say is that when booting a game, wxWidgets is used to create the window.

https://github.com/DHrpcs3/rpcs3/blob/31bf130b1c7acf78ef334a2bdb0550971e6d2821/rpcs3/Gui/GLGSFrame.cpp#L11

Furthermore, wxWidgets is responsible for creating the OpenGL context to use with this window. The version of wxWidgets  being used doesn't allow for any configurability over the version of OpenGL. It would seem that this doesn't affect anyone on windows but for me, running on OSX, it looks like a version 2.1 context is created instead of the highest supported version for OSX, 4.1.

So I guess you can boil the purpose of creating this issue into three questions. 

First, does my logic for how the OpenGL context gets created sound correct? Secondly, Is this what is expected or should this be addressed? Finally, are there any other OSX users noticing this (assuming there are some for that matter)?

**Comment by Syphurith** [comment]
Seems OSX version also limits the support ([here](http://stackoverflow.com/questions/22922367/loading-an-opengl-4-1-context-on-osx)). Unfortunately i can not access apple website to find more clues. Someone made [patch](https://hg.libsdl.org/SDL/rev/9bd65b58278e) for SDL library, not sure if there is any relation.
You have several ways to handle the problem:
* Check your OSX version and try to upgrade it. Well iirc since Apple limit those older hardware from fetching the upgrades, you may even need another one. Indeed i have already heard of the story that many OSX users in China installed Windows after first use. Or simply you can think of installing Win/Linux, and keep it dual-boot.
* If OSX version does support it, find any way to patch the libraries. With a glance of those search result this doesn't appear to be an easy job. Damn Apple.
* Or update the wxwidgets (wxGLContext?). Yep i saw [patch](http://trac.wxwidgets.org/ticket/16402) cause of rpcs3 there (7 months ago).
* If you know how to emulate the opeartions, and you can implement the functions in another way, thus you can downgrade the requirements. To me i know `IE7.js` which let IE6/7/8 to emulate the higher version of IE.
* Contact Apple for technical support. You can ask for a technical 'how to'.

Finally, good luck, whichever you choose..

**Comment by uyjulian** [comment]
Mac OS X requires you to use core context to enable 3.0+ features.
Mac OS X doesn't support compatibility context with 3.0+ feature.

**Comment by tambry** [comment]
@uyjulian So is it right that it's basically Mac OS X's fault and since it isn't our problem, this issue can be closed?

**Comment by uyjulian** [comment]
@tambry If you are only using 3.0+ OpenGL features, you need to explicitly create an core context.
If you are using (2.1 and below) and 3.0+ OpenGL features, it's OSX's fault.
If you are using (2.1 and below) OpenGL features, you don't need to do anything.

**Comment by tambry** [comment]
@uyjulian I think we use both - from 2.1 and 3.0+ OpenGL features. Might be nice if another developer could confirm it though.

**Comment by sturmE** [comment]
@uyjulian wxWidgets seems to be responsible for creating the context and the current version being used does not allow for any configuration options to be passed in so it is creating the compatibility context targeting 2.1. 

According to this [documentation](http://docs.wxwidgets.org/trunk/glcanvas_8h.html#aae05225933a42f81e7c4a9fb286596f9aa75a4e738b4fa60c91409090f40d7a40), newer versions of wxWidgets allow you to request a core context.

**Comment by Syphurith** [comment]
@sturmE well then, you can try replace your wxwidget version, and create a core context with it.

**Comment by tambry** [comment]
@sturmE Is this fixed? Since we updated our wxWidgets version a while ago.

**Comment by Bigpet** [comment]
@tambry  we didn't change this, we still create a default context here: https://github.com/RPCS3/rpcs3/blob/5ae6586803fdcc5594ec83df4510bd2723c4c1c1/rpcs3/Gui/GLGSFrame.cpp#L12

and we use all kinds of deprecated and old stuff that prevents us from just using a core-context, like this for example: https://github.com/RPCS3/rpcs3/blob/faf0a0d851db4882567d314c7216ac46c524a586/rpcs3/Emu/RSX/GL/GLGSRender.cpp#L2110 (I mean that's one example that you can easily change, but there's some others that are not that easy)

**Comment by Aaahh** [comment]
Well then, lets change them? 
Sounds like a lot of work but eh, I sometimes have free time.

**Comment by Bigpet** [comment]
@Aaahh well it might not be that much work (I did it as a temporary change locally a year ago or so to make it run with NSight), but I doubt it will help with OSX support.

Apple still only supports OpenGL 4.1 (and doesn't show much of any hurry to update versions since they are more focused on metal) and the shader it complains about is written with the GLSL version belonging to OpenGL 4.2+ in mind.

**Comment by Aaahh** [comment]
@Bigpet what about performance? OpenGl 4.2 removed glstart and abunch of other things didn't it? I bet apple uses opengl 4.1 because they still use all those.

**Comment by Bigpet** [comment]
@Aaahh no buddy, any core context 3.0+ deprecated immediate mode

**Comment by Aaahh** [comment]
@Bigpet yeah but they could still use deprecated methods? 
Whatever, I'm trying to find something to do on this project.

**Comment by Bigpet** [comment]
@Aaahh no, *3.0 core contexts* could not use deprecated methods, that was the whole point of core contexts

**Comment by Aaahh** [comment]
@Bigpet yeah that makes sense, I guess I just derped

**Comment by vlj** [comment]
The core/compatibility context thing was written to simplify drivers. It is allowed by Khronos to support only core context and GL 3.1+ version. Beforehand a driver had to support everything to claim GL 2.0 or 2.1 conformance despite gfx card architecture being very different from the GL 1.0 age.

Mesa and Apple decided not to ship a compat context to simplify drivers development, while Nvidia and AMD have one (likely to fit professional software needs).

I don't know what make rpcs 3 require GL 4.2 atm. Like bigpet said it's likely possible to use GL 4.1.
However performance wise it will likely lag behind DX12 backend... 
To get correct performance a gfx backend must have a low overhead, and thus avoid as many state/resource creation as possible. 
This means for instance avoiding creating vertex array object and reuse them as much as possible, which is only possible with ARB_vertex_attrib_binding in GL 4.3 which provide an API splitting vertex specification and actual data storage.
Another exemple is avoiding recreating/copying vertex/index buffers per frame ; thus glBufferSubData and classic glMapRange are not good candidate there. Maybe round robing buffer algorithm in driver may prevent actual buffer recreation but it's rather unreliable. A good way to handle this is to use persistent buffer (available in GL 4.4)  and use a ring buffer pattern (like I do with DX12). Plus it may avoid an extra copy since it allows to directly write data in memory.
And there is the Direct State Access extension, available in GL 4.5, that avoids binding object nearly all the time and thus provide lower overhead.

TL;DR while rpcs3 may use GL 4.1 in its current form but it's unlinkely it will provide decent performance. It probably requires GL 4.5 to get good fps which I fear will never make it to OSX. Or use Metal.

**Comment by tambry** [comment]
Can this be closed?
